p4include/psa.p4

```diff
@@ -0,0 +1,675 @@
+    void reset        (in S index);
+    void start        (in S index);
+    void stop         (in S index);
+  }
+  */
+}
+
+/// DirectCounter: RFC-XXXX
+extern DirectCounter<W> {
+  DirectCounter(entry_key key, CounterType_t counter_type);
+  void count();
+
+  /*
+  @ControlPlaneAPI
+  {
+    W    read<W>      (in entry_key key);
+    W    sync_read<W> (in entry_key key);
+    void set          (in W seed);
+    void reset        (in entry_key key);
+    void start        (in entry_key key);
+    void stop         (in entry_key key);
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Meters: RFC 2698
+ *
+ * Meters are a more complex mechanism for keeping statistics about
+ * the packets that trigger a table. The meters specified in the PSA
+ * are 3-color meters.
+ */
+
+/// Meter types
+enum MeterType_t {
+    packets,
+    bytes
+}
+/// Meter colors
+enum MeterColor_t { RED, GREEN, YELLOW };
+
+/// Meter
+extern Meter<S> {
+  Meter(S n_meters, MeterType_t type);
+  MeterColor_t execute(in S index, in MeterColor_t color);
+
+  /*
+  @ControlPlaneAPI
+  {
+    reset(in MeterColor_t color);
+    setParams(in S committedRate, out S committedBurstSize
+              in S peakRate, in S peakBurstSize);
+    getParams(out S committedRate, out S committedBurstSize
+              out S peakRate, out S peakBurstSize);
+  }
+  */
+}
+
+/// DirectMeter
+extern DirectMeter {
+  DirectMeter(MeterType_t type);
+  MeterColor_t execute(in MeterColor_t color);
+
+  /*
+  @ControlPlaneAPI
+  {
+    reset(in MeterColor_t color);
+    void setParams<S>(in S committedRate, out S committedBurstSize
+                      in S peakRate, in S peakBurstSize);
+    void getParams<S>(out S committedRate, out S committedBurstSize
+                      out S peakRate, out S peakBurstSize);
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Registers
+ *
+ * Registers are stateful memories whose values can be read and
+ * written in actions. Registers are similar to counters, but can be
+ * used in a more general way to keep state.
+ *
+ * Although registers cannot be used directly in matching, they may be
+ * used as the RHS of an assignment operation, allowing the current
+ * value of the register to be copied into metadata and be available
+ * for matching in subsquent tables.
+ */
+extern Register<T, S> {
+  Register(S size);
+  T    read  (in S index);
+  void write (in S index, in T value);
+
+  /*
+  @ControlPlaneAPI
+  {
+    T    read<T>      (in S index);
+    void set          (in S index, in T seed);
+    void reset        (in S index);
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Random
+ *
+ * The random extern provides a reliable, target specific number generator
+ * in the min .. max range.
+ */
+
+/// The set of distributions supported by the Random extern.
+enum RandomDistribution {
+  PRNG,
```

Comment:
A uniform random generator is a useful thing on its own.

A uniform random generator plus some formulas using + - << and if's is enough to implement piecewise linear approximations of other generators.

A uniform random generator _plus_ a lookup table with an exact key that is a uniform random value enables one to implement whatever distribution you can imagine, if you are willing to spend the table's storage to do it (and it can even be modified by the control plane at run time).

Given all of those choices, and given that a high performance target would probably implement non-uniform distributions using one of the methods above, I would think that uniform random would be sufficient, and probably all that some targets would be willing to implement anyway.